### PR TITLE
feat(tokens): adds support for creating tokens to the sdk

### PR DESCRIPTION
**Pull request by jakekgrog**
# Why
Need to support the creation of tokens in the PHP SDK

# How
- Add `createClientSideDecryptToken()` to allow

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -151,6 +151,38 @@ $evervault->createRunToken($functionName = string, $data = array or object)
 | `$functionName` | `string` | Name of the Function the Run Token should be created for |
 | `$data`      | `array` or `object` | Payload that the Run Token can be used with. This is an optional parameter. If not provided or the payload is an empty object, the Run Token will be valid for any payload. |
 
+### $evervault->decrypt()
```
Comment (jeremywagemans) [inline README.md:154]: this section is duplicated!

Comment (jakekgrog) [reply README.md:154]: how'd that get there lol

File: src/Evervault/EvervaultHttp.php [hunk 2]
```diff
@@ -185,4 +259,16 @@ public function decrypt($data) {
...
     }
+
+    public function createToken($action, $data, $expiry) {
+        $payload = array(
+            'action' => $action,
+            'payload' => $data,
+            'expiry' => $expiry,
+        );
+
+        $response = $this->_makeCreateTokenRequest($payload);
```
Comment (jeremywagemans) [inline src/Evervault/EvervaultHttp.php:270]: This is not needed. You can just rely on `_makeApiRequest`, similary to how it's done in `decrypt`

File: README.md [hunk 3]
```diff
@@ -151,6 +151,38 @@ $evervault->createRunToken($functionName = string, $data = array or object)
...
+
+| Parameter | Type  | Description          |
+| --------- | ----- | -------------------- |
+| encrypted | Array | Data to be decrypted |
+
+### $evervault->createClientSideDecryptToken()
+
+`$evervault->createClientSideDecryptToken()` creates a time-bound token that can be used to decrypt previously encrypted data.
+
+If the `$data` parameter is provided, the token can only be used to decrypt that specific payload. Otherwise, the token can be used to decrypt any payload.
```
Comment (jeremywagemans) [inline README.md:158]: for decrypt, it is actually required to pass the payload :)

**Review by jeremywagemans — CHANGES_REQUESTED** [review]
Added a few comments!

File: src/Evervault/EvervaultHttp.php [hunk 4]
```diff
@@ -169,6 +178,72 @@ private function _makefunctionRunRequest($functionName, $body = [], $headers = [
         return $this->_handleApiResponse($this->curl, $response, $headers);
     }
 
+    private function _makeCreateTokenRequest($payload, $headers=[]) {
```
Comment (jeremywagemans) [inline src/Evervault/EvervaultHttp.php:181]: Remove this

File: src/Evervault/EvervaultHttp.php [hunk 5]
```diff
@@ -169,6 +178,72 @@ private function _makefunctionRunRequest($functionName, $body = [], $headers = [
...
+            CURLOPT_RETURNTRANSFER,
+            true
+        );
+
+        $response = curl_exec($this->curl);
+
+        return $this->_handleApiResponse($this->curl, $response);
+    }
+
+    private function _makeDecryptRequest($data, $headers=[]) {
```
Comment (jeremywagemans) [inline src/Evervault/EvervaultHttp.php:214]: Remove this

File: src/Evervault/Evervault.php [hunk 6]
```diff
@@ -80,6 +80,18 @@ public function decrypt($data) {
         return $this->httpClient->decrypt($data);
     }
 
+    public function createClientSideDecryptToken($data, $expiry) {
+        if (!$data) {
+            throw new EvervaultError('Please provide a payload');
```
Comment (jeremywagemans) [inline src/Evervault/Evervault.php:85]: Can you improve this error message to make it more user friendly?

```suggestion
            throw new EvervaultError('The `$data` parameter is required and ensures the issued token can only be used to decrypt that specific payload.');
```

File: src/Evervault/EvervaultHttp.php [hunk 7]
```diff
@@ -185,4 +194,16 @@ public function decrypt($data) {
...
     }
+
+    public function createToken($action, $data, $expiry) {
+        $payload = array(
+            'action' => $action,
+            'payload' => $data,
+            'expiry' => $expiry,
+        );
+
+        $response = $this->_makeApiRequest('POST', $this->$createTokenPath, $payload, [], true);
```
Comment (jeremywagemans) [inline src/Evervault/EvervaultHttp.php:205]: This doesn't work:

```suggestion
        $response = $this->_makeApiRequest('POST', $this->createTokenPath, $payload, [], true);
```

File: src/Evervault/Evervault.php [hunk 8]
```diff
@@ -80,6 +80,18 @@ public function decrypt($data) {
...
+        }
+
+        if ($expiry) {
+            $expiry = $expiry * 1000;
+        }
+        
+        return $this->httpClient->createToken("api:decrypt", $data, $expiry);
```
Comment (jeremywagemans) [inline src/Evervault/Evervault.php:92]: The grant is invalid:

```suggestion
        return $this->httpClient->createToken("decrypt:api", $data, $expiry);
```

File: src/Evervault/Evervault.php [hunk 9]
```diff
@@ -80,6 +80,18 @@ public function decrypt($data) {
         return $this->httpClient->decrypt($data);
     }
 
+    public function createClientSideDecryptToken($data, $expiry) {
```
Comment (jeremywagemans) [inline src/Evervault/Evervault.php:83]: We should make this parameter optional

```suggestion
    public function createClientSideDecryptToken($data, $expiry = null) {
```

**Review by jeremywagemans — CHANGES_REQUESTED** [review]
Some implementation issues

File: src/Evervault/EvervaultHttp.php [hunk 10]
```diff
@@ -96,7 +96,7 @@ private function _handleApiResponse($curl, $response, $headers = []) {
             throw new EvervaultError('Your API key was invalid. Please verify it matches your API key in the Evervault Dashboard.');
         } else if ($responseCode === 403) {
             throw new EvervaultError('Your API key does not have the required permissions to perform this action. You can update your API key permissions in the Evervault Dashboard.');
-        } else if ($responseCode !== 200) {
+        } else if ($responseCode !== 200 || $responseCode !== 201) {
```
Comment (jeremywagemans) [inline src/Evervault/EvervaultHttp.php:99]: This should be `responseCode !== 200 && $responseCode !== 201`.

**Review by jeremywagemans — CHANGES_REQUESTED** [review]
Looks good except one last issue!

**Review by jeremywagemans — APPROVED** [review]
lgtm!

**Pull request merged**